### ISSUE TITLE
CmdStanR backend? 

### ISSUE THREAD

**Issue by jgabry** [facebook/prophet#1552]
It's great that prophet allows CmdStanPy as a backend, so I wanted give you an update and let you know that on the R side CmdStanR is now finally also available (https://github.com/stan-dev/cmdstanr). We haven't done a CRAN release yet but the interface is stable enough now that people are using it and other packages (e.g. brms) are already starting to offer CmdStanR as an optional backend. 

If you're interested in offering a CmdStanR backend definitely let us (me or @rok-cesnovar) know if you have any questions or run into any trouble.

**Comment by bletham** [comment]
This is great to hear! We have had problems with rstan in the past where passing large matrices back from Stan to R has noticeable overhead (it can be several seconds), so I'll be interested to see if CmdStanR is an improvement on that. I think we could closely parallel the refactoring done to add the CmdStanPy backend to add support. For reference, #1045 was the issue for the CmdStanPy interface.

**Comment by tcuongd** [comment]
Hey I'm keen to give this a go :) I'm not too sure what the best way to go about this is though. It seems like the approach we currently take is to build the C++ code for the stan model on install, then save it as a C++ module in the `stanmodels` object that can then be called in source code (ref: https://github.com/facebook/prophet/blob/master/R/R/stanmodels.R).

I'm not sure if `cmdstanr` can work directly from C++ code (just looking at this: https://mc-stan.org/cmdstanr/reference/cmdstan_model.html), so would we need another way to compile and save the stan model on install? Usually `cmdstan_model()` would recognize if the executable is up to date, so we could compile on install then call `cmdstan_model("inst/prophet.stan")` to instantiate the model in the code, but I'm not sure if that's good practice. Unless we can save the model as an RDS object instead?

Would appreciate any guidance on this!

**Comment by tcuongd** [comment]
I've made a start on this here: https://github.com/facebook/prophet/pull/1806
I've just added functionality in the code to switch between the two backends for `fit.prophet` so far, but still not sure of the best way to handle installation and initial compilation.

How it's currently done in that draft PR assumes that `rstan` will be the default backend and will be installed regardless, but the user has the option to choose `cmdstanr` to fit if they have it installed. Once `cmdstanr` is available on CRAN perhaps we'll be able to remove the `rstan` and `Rcpp` dependencies and use `cmdstanr` as the default?
